src/coordinator/functions/temporal/base.go

```diff
@@ -0,0 +1,348 @@
+// 2. For the current block, figure out whether we have enough previous blocks which can help process it
+// 3. For the blocks after current block, figure out which can be processed right now
+// 4. Process all valid blocks from #3, #4 and mark them as processed
+// 5. Run a sweep face to free up blocks which are no longer needed to be cached
+func (c *baseNode) Process(ID parser.NodeID, b block.Block) error {
+	iter, err := b.StepIter()
+	if err != nil {
+		return err
+	}
+
+	meta := iter.Meta()
+	bounds := meta.Bounds
+	blockDuration := bounds.Duration
+	// Figure out the maximum blocks needed for the temporal function
+	maxBlocks := int(math.Ceil(float64(c.op.duration) / float64(blockDuration)))
+
+	// Figure out the leftmost block and rightmost blocks
+	leftRangeStart := bounds.Previous(maxBlocks).Start
+	queryStartBounds := bounds.Nearest(c.transformOpts.TimeSpec.Start)
+
+	if leftRangeStart.Before(queryStartBounds.Start) {
+		leftRangeStart = queryStartBounds.Start
+	}
+
+	rightRangeStart := bounds.Next(maxBlocks).Start
+	queryEndBounds := bounds.Nearest(c.transformOpts.TimeSpec.End.Add(-1 * bounds.StepSize))
+
+	if rightRangeStart.After(queryEndBounds.Start) {
+		rightRangeStart = queryEndBounds.Start
+	}
+
+	// Process left side of the range
+	leftBlks, emptyLeftBlocks := c.processLeft(b, bounds, maxBlocks, leftRangeStart)
+
+	processRequests := make([]processRequest, 0, len(leftBlks))
+	// If we have all blocks for the left range in the cache, then process the current block
+	if !emptyLeftBlocks {
+		processRequests = append(processRequests, processRequest{blk: b, deps: leftBlks, bounds: bounds})
+	}
+
+	leftBlks = append(leftBlks, b)
+
+	// Process right side of the range
+	rightBlks, emptyRightBlocks := c.processRight(b, bounds, maxBlocks, rightRangeStart)
+
+	for i := 0; i < len(rightBlks); i++ {
+		lStart := maxBlocks - i
+		if lStart > len(leftBlks) {
+			continue
+		}
+
+		deps := leftBlks[len(leftBlks)-lStart:]
+		deps = append(deps, rightBlks[:i]...)
+		processRequests = append(processRequests, processRequest{blk: rightBlks[i], deps: deps, bounds: bounds.Next(i + 1)})
+
+	}
+
+	// If either the left range or right range wasn't fully processed then cache the current block
+	if emptyLeftBlocks || emptyRightBlocks {
+		if err := c.cache.Add(bounds.Start, b); err != nil {
+			return err
+		}
+	}
+
+	return c.processCompletedBlocks(processRequests, queryStartBounds, queryEndBounds, maxBlocks)
+}
+
+// processLeft processes the current block. For the current block, figure out whether we have enough previous blocks which can help process it
+func (c *baseNode) processLeft(b block.Block, bounds block.Bounds, maxBlocks int, leftRangeStart time.Time) ([]block.Block, bool) {
+	leftRangeTimes := make([]time.Time, 0, maxBlocks)
+	for t := leftRangeStart; t.Before(bounds.Start); t = t.Add(bounds.Duration) {
+		leftRangeTimes = append(leftRangeTimes, t)
+	}
+
+	leftBlks := c.cache.MultiGet(leftRangeTimes)
```

Comment:
Would it be better to push some logic down to `MultiGet`, where it shortcircuits out on a missing cache value, then returns the partial list and `false`, rather than doing additional processing here and in `processRight`?

Reply:
yeah sounds good.